fix bootstrap-salt.sh for FreeBSD

kgbsd opened the pull request:
FreeBSD has py36-salt packages which handle all dependencies, etc - simple as pkg install -y py36-salt and then enabling it.  Don't need dependencies or git hacks.

### What does this PR do?
able to bootstrap salt on FreeBSD

### What issues does this PR fix or reference?
able to bootstrap salt on FreeBSD

### Previous Behavior
bootstrap FreeBSD did not work, failed with logic and dependency hell

### New Behavior
installs FreeBSD built py36-salt package with correct paths, etc.

kgbsd committed:
fix bootstrap-salt.sh for FreeBSD

FreeBSD has py36-salt packages which handle all dependencies, etc - simple as pkg install -y py36-salt and then enabling it.  Don't need dependencies or git hacks.

kgbsd commented:
https://github.com/saltstack/salt-bootstrap/issues/1377

bryceml commented on bootstrap-salt.sh:
```diff
@@ -238,7 +238,11 @@ _ECHO_DEBUG=${BS_ECHO_DEBUG:-$BS_FALSE}
 _PIP_ALL=${BS_PIP_ALL:-$BS_FALSE}
-_SALT_ETC_DIR=${BS_SALT_ETC_DIR:-/etc/salt}
+if uname -a | grep FreeBSD > /dev/null; then
+  SALT_ETC_DIR=${BS_SALT_ETC_DIR:-/usr/local/etc/salt}
```
are you missing a `_` here?

bryceml commented:
The git installation option is not optional to implement.  It allows for installing from any arbitrary hash/branch/tag from the salt repo.  That functionality can't go away.

kgbsd committed:
Update bootstrap-salt.sh

vutny commented on bootstrap-salt.sh:
```diff
@@ -238,7 +238,11 @@ _ECHO_DEBUG=${BS_ECHO_DEBUG:-$BS_FALSE}
 _PIP_ALL=${BS_PIP_ALL:-$BS_FALSE}
-_SALT_ETC_DIR=${BS_SALT_ETC_DIR:-/etc/salt}
+if uname -a | grep FreeBSD > /dev/null; then
+  _SALT_ETC_DIR=${BS_SALT_ETC_DIR:-/usr/local/etc/salt}
```
There are different indentation style of 2 spaces instead of 4 in all lines you have changed.

vutny reviewed (changes requested):
I guess, installing OS distro supplied package (or port) is OK when no arguments given to the bootstrap script, or the `latest`. Probably, it would be complicated to support earlier versions, but there could be a an error message about that.
If git installations are badly broken, I think it should be addressed separately.

At very least, you could just use one-liner specific to your OS to install available packages.

kgbsd committed:
Merge branch 'develop' into patch-1

kgbsd committed:
Merge branch 'develop' into patch-1

kgbsd committed:
working on FreeBSD

kgbsd commented:
pulled latest from git
Fixed indentation back to 4
merged my changes, ran this on FreeBSD 12, and it installed, and properly got salt minion running.
I cleaned up a lot of [really old] hacks, which partially were just wrong, and partially are not needed. This makes the FreeBSD part much cleaner, simpler and KISS compliant.
This is 100% better than having a non working script, I haven't had time to test install from git or other FreeBSD versions. I don't see why other version would fail.
With FreeBSD and pkg stuff just works, no dependencies to manually sort.

vutny commented on bootstrap-salt.sh:
```diff
@@ -5268,113 +5272,22 @@ __freebsd_get_packagesite() {
 }
 
 install_freebsd_git_deps() {
     install_freebsd_9_stable_deps || return 1
```
Looks like this function was replaced with `install_freebsd_deps()`: 
```suggestion
    install_freebsd_stable_deps || return 1
```

kgbsd committed:
Update bootstrap-salt.sh

missed a freebsd_9_stable_deps part

Co-Authored-By: Denys Havrysh <Denys.Gavrysh@gmail.com>

kgbsd committed:
Update bootstrap-salt.sh

vutny commented on bootstrap-salt.sh:
```diff
@@ -238,7 +238,11 @@ _ECHO_DEBUG=${BS_ECHO_DEBUG:-$BS_FALSE}
+    _SALT_ETC_DIR=${BS_SALT_ETC_DIR:-/usr/local/etc/salt}
+else
+    _SALT_ETC_DIR=${BS_SALT_ETC_DIR:-/etc/salt}
+fi
```
I think overriding default values for `_SALT_ETC_DIR` and `_PKI_DIR` should be done later, let it be in `__configure_freebsd_pkg_details()` function, for the sake of keeping related code close to where it would be used.
So, no need for an `if-else` statement here, it would look much cleaner.

vutny commented on bootstrap-salt.sh:
```diff
@@ -5268,130 +5272,31 @@ __freebsd_get_packagesite() {
-
-install_freebsd_11_stable_deps() {
-    install_freebsd_9_stable_deps
+    echo 'not hijacking pkg repo configurations'
```
This functions appears to be dummy and not being referenced at all.
I suggest it would be a perfect place for overriding `_SALT_ETC_DIR` and `_PKI_DIR` variables.
Also, it should be sourced from `install_freebsd_deps()`.

vutny commented on bootstrap-salt.sh:
```diff
@@ -5268,130 +5272,31 @@ __freebsd_get_packagesite() {
-    SALT_DEPENDENCIES=$(/usr/local/sbin/pkg search ${FROM_FREEBSD} -R -d sysutils/py-salt | grep -i origin | sed -e 's/^[[:space:]]*//' | tail -n +2 | awk -F\" '{print $2}' | tr '\n' ' ')
+    SALT_DEPENDENCIES=$(/usr/local/sbin/pkg search -R -d sysutils/py-salt | grep -i origin | sed -e 's/^[[:space:]]*//' | tail -n +2 | awk -F\" '{print $2}' | tr '\n' ' ')
+    SALT_DEPENDENCIES=$(/usr/local/sbin/pkg search -R -d py36-salt | grep 'origin:' \
+        | tail -n +2 | awk -F\" '{print $2}' | sed 's#.*/py-#py36-#g')
```
The latter line will override the value set in the previous one. Is there something missed?

vutny reviewed (changes requested):
Thanks for addressing comments and keep working on this, @kgbsd !

I just have few small inquires to make up the changes here. Looks good overall.

kgbsd committed:
update script per comments / review

move FreeBSD specifics to be under 'FreeBSD Install Functions'
remove old code that was left in there just for reference while debugging

kgbsd committed:
give py-requests pkg proper name

kgbsd committed:
cannot use 'if not set' expression here

_SALT_ETC_DIR=${BS_SALT_ETC_DIR:-/usr/local/etc/salt}
Says if variable is not set, set it to that.  By the time the script gets this far down, the variable is already set, so have to hardcode it to be /usr/local/etc/salt - part of the reason I had an 'if' statement around line 242, before the variable is set to default.

vutny commented on bootstrap-salt.sh:
```diff
@@ -5268,130 +5268,30 @@ __freebsd_get_packagesite() {
-install_freebsd_12_stable_deps() {
-    install_freebsd_9_stable_deps
+install_freebsd_deps() {
+    __configure_freebsd_pkg_details()
```
This is a syntax error, I guess just some copy-pasting mistake. Remove parentheses.
```suggestion
    __configure_freebsd_pkg_details
```

kgbsd committed:
fix copy paste error

remove () in a call to shell function.

Co-Authored-By: Denys Havrysh <Denys.Gavrysh@gmail.com>

vutny reviewed (changes requested):
This looks much better now, thanks for your dedication @kgbsd !

We're approaching to the final cut: the function `__freebsd_get_packagesite()` has got unused.
I think you want to drop it as well.

kgbsd committed:
Update bootstrap-salt.sh

vutny commented on bootstrap-salt.sh:
```diff
@@ -5509,34 +5361,18 @@ install_freebsd_9_stable_post() {
 install_freebsd_git_post() {
     if [ -f $salt_conf_file ]; then
         rm -f $salt_conf_file
     fi
```
We're close!
It seems that the `salt_conf_file` variable is not set anymore. That will cause an error. Let's drop the whole `if` statement here.

kgbsd committed:
salt_conf_file unused ${salt_conf_file} var

${salt_conf_file} no longer needed, this conf file for saltstack repo which is not used.

vutny commented:
@kgbsd We have merged a fix just a few moments ago for CI to pass.
Please rebase on latest `develop`, let's look at the build status and we will make this through.
Thanks!

kgbsd committed:
Merge branch 'develop' into patch-1

vutny commented:
@s0undt3ch The build is pending here after the rebase, would you mind to "APPROVE" it on the Drone? Thanks!

s0undt3ch commented:
Done

vutny reviewed (approved):
Good job, @kgbsd !

Last CI run has passed with the exception of flaky Opensuse test, which could be ignored.
Although the branch became even with `develop` (unfortunately again), I think it is well done to be merged from the general code quality perspective and contains no regressions. Would you mind to rebase once more just for consistency?

I don't have access to the resources to provision FreeBSD machine for now, but it would be great to do a basic testing for this change. Could you please give it a shot, @Ch3LL ?

Anyway, I think this is good to go, @s0undt3ch .

kgbsd committed:
Merge branch 'develop' into patch-1

vutny commented:
The build has stuck for some reason... But anyway, @s0undt3ch , this is good and could be merged.

vutny commented:
:tada:
